T19437

smspillaz opened the pull request:
Allow distributors to override the refspec to update from
    
This can effectively be used to implement "checkpoint" epochs, by
placing a new refspec in /usr/share/eos-updater/upgrade-refspec .
For instance, a distributor might wish to ensure that a client is
at least running version X before attempting to pull and apply
version Y. To do that, they would make a final commit on X which
sets /usr/share/eos-updater/upgrade-refspec to a new refspec
that the updater should look at the next time it pulls and deploys
an OSTree when it is booted into the system on that final commit.

This ensures that for systems not yet booted into that final commit,
the updater will only pull revisions up to the final one and will
only begin to pull new revisions on the new refspec once booted
into the system on the final commit.

This PR subsumes T16682, which gives us a mechanism to install
flatpaks upon reboot of the system.

https://phabricator.endlessm.com/T19437

pwithnall commented on tests/eos-test-utils.c:
```diff
@@ -603,7 +603,10 @@ prepare_commit (GFile *repo,
     if (g_file_query_exists (commit_file, NULL))
-      return TRUE;
+      {
+        if (out_checksum != NULL)
```
This doesn’t `return TRUE` any more if `(out_checksum == NULL)`.

pwithnall commented on tests/eos-test-utils.c:
```diff
@@ -773,6 +773,33 @@ generate_delta_files (GFile *repo,
+            return key;
+        }
+    }
+  while (wanted_commit_number > 0);
```
It would be clearer to restructure this as a `for` loop.

pwithnall commented on tests/eos-test-utils.c:
```diff
@@ -773,6 +773,33 @@ generate_delta_files (GFile *repo,
 }
 
+static const gchar *
+get_last_ref (GHashTable *ref_to_commit,
```
This function could do with a documentation comment.

pwithnall commented on tests/eos-test-utils.c:
```diff
@@ -847,6 +847,10 @@ update_commits (EosTestSubserver *subserver,
         return FALSE;
 
+      /* No checksum means that we've already written out this commit's ref */
+      if (!checksum)
```
Be explicit in `NULL` comparisons please: `if (checksum == NULL)`.

pwithnall commented on tests/Makefile.am:
```diff
@@ -148,6 +149,12 @@ test_update_broken_delta_LDFLAGS = $(test_ldflags)
+test_update_refspec_checkpoint_CFLAGS = $(test_cflags)
+test_update_refspec_checkpoint_LDFLAGS = $(test_ldflags)
+test_update_refspec_checkpoint_LDADD = $(test_ldadd)
+test_update_refspec_checkpoint_SOURCES = test-update-refspec-checkpoint.c
```
This new file seems to be missing from the commit.

smspillaz replied:
Yikes, thanks for catching. I'll add it now.

pwithnall commented on src/eos-updater-poll-common.c:
```diff
@@ -1183,7 +1255,7 @@ run_fetchers (EosMetadataFetchData *fetch_data,
 
       /* Send metrics about our ref: this is the ref we’re going to upgrade to,
        * but that’s always the same as the one we’re currently on. */
```
This comment is now out of date.

pwithnall commented on src/eos-updater-poll-common.c:
```diff
@@ -303,6 +303,78 @@ get_booted_refspec (gchar **booted_refspec,
+                                 error);
+    }
+
+  if (!ostree_parse_refspec (refspec, &remote, &ref, error))
```
Might want to `g_strstrip (refspec)` first, for robustness. (If someone edits the `upgrade-refspec` file with a text editor, it will probably automatically append a `\n`.)

pwithnall commented:
@smspillaz: When you push commits which fix review issues, can you please create them as fixup commits (`git commit --fixup`) so that we don’t end up with >200 commits on a branch and a commit history which is a) impossible to bisect sensibly afterwards; and b) impossible to post-hoc squash down easily?

smspillaz commented:
> @smspillaz: When you push commits which fix review issues, can you please create them as fixup commits (git commit --fixup) so that we don’t end up with >200 commits on a branch and a commit history which is a) impossible to bisect sensibly afterwards; and b) impossible to post-hoc squash down easily?

Ah nice, I didn't know about that feature - I'll make use of in in the future. In any event, I'm happy to squash this branch down into something sensible if we're ready to merge it. Are there any further things to address?

pwithnall commented on tests/eos-test-utils.c:
```diff
@@ -775,6 +775,18 @@ generate_delta_files (GFile *repo,
 
+/**
+ * get_last_ref:
+ * @ref_to_commit: a #GHashTable mapping refs to commit numbers.
```
Nitpick: Even though `g-ir-scanner` will never see this static method, I think it’s good to get into the habit of just adding GObject introspection annotations in all doc comments. So this should have `(element-type utf8 utf8)` (check that).

pwithnall commented on tests/eos-test-utils.c:
```diff
@@ -775,6 +775,18 @@ generate_delta_files (GFile *repo,
+ * @ref_to_commit: a #GHashTable mapping refs to commit numbers.
+ * @wanted_commit_number: the commit number to count down from.
+ *
+ * Look through the ref_to_commit hashtable (which usually maps refs to
```
‘usually’?

pwithnall commented on tests/test-update-refspec-checkpoint.c:
```diff
@@ -0,0 +1,238 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright © 2016 Kinvolk GmbH
```
Really? How much of this is Kinvolk’s code and how much is ours?

pwithnall commented on tests/test-update-refspec-checkpoint.c:
```diff
@@ -0,0 +1,238 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * Authors:
+ *  - Krzesimir Nowak <krzesimir@kinvolk.io>
```
And here.

pwithnall commented on tests/test-update-refspec-checkpoint.c:
```diff
@@ -0,0 +1,238 @@
+}
+
+static void
+test_update_refspec_checkpoint (EosUpdaterFixture *fixture,
```
As with all test cases, please add a comment above the test case function which explains roughly what it tries to test and how it goes about that.

pwithnall commented on src/eos-updater-poll-common.c:
```diff
@@ -76,7 +76,9 @@ G_STATIC_ASSERT (G_N_ELEMENTS (order_key_str) == EOS_UPDATER_DOWNLOAD_LAST + 1);
 
-static const gchar *const REFSPEC_TO_UPGRADE_ON_PATH = PKGDATADIR "/upgrade-refspec";
+/* This path looks hardcoded, but it is actually a relative path, relative to
+ * the current OSTree deployment and not the rootfs. */
```
What’s the reasoning for this change? If we decide to change the `PREFIX` for eos-updater, surely we want it to look in a correspondingly different subpath of the deployment?

pwithnall commented on tests/test-update-refspec-checkpoint.c:
```diff
@@ -0,0 +1,238 @@
+  g_assert_no_error (error);
+
+  /* Now update the client. We stopped making commits on this
+   * ref, so it is effectively a "checkpoint" and we hsould only have
```
Nitpick: typo ‘should’

pwithnall commented on tests/test-update-refspec-checkpoint.c:
```diff
@@ -0,0 +1,238 @@
+const gchar *next_refspec = "REMOTE:REFv2";
+
+static void
+following_refspec_files (guint         commit,
```
This function could do with a brief comment saying what it does too. The function name is not very obvious to me.

pwithnall commented on src/eos-updater-poll-common.c:
```diff
@@ -305,14 +301,45 @@ get_booted_refspec (OstreeDeployment  *booted_deployment,
+  /* Look up the refspec to upgrade from here */
+  metadata = g_variant_get_child_value (commit, 0);
+  ref_for_deployment_variant = g_variant_lookup_value (metadata,
+                                                       "eos.upgrade-ref-for-deployment",
```
What happens if `eos.upgrade-ref-for-deployment` and `endoflife-rebase` are both set to conflicting things? As I said on Slack, I think it would be better if this was a boolean ‘boot into this one before continuing’, used in conjunction with `endoflife-rebase` which provides the ref name to switch to.

smspillaz replied:
Ehhh. I'm not sure if we can meaningfully do that without having to do a bunch of gymnastics in stepping back through history in the pull to find the relevant commit. We control the ostree build process so is `eos.upgrade-ref-for-deployment` and `endoflife-rebase` are set to conflicting things on the *same* commit, then to be honest that's the ostree builder's problem.

pwithnall replied:
> if `eos.upgrade-ref-for-deployment` and `endoflife-rebase` are set to conflicting things on the same commit, then to be honest that's the ostree builder's problem.

It’s definitely always better to prevent such situations ever being possible through the design of data structures, rather than having to remember to prohibit them everywhere that data structure is populated (which might eventually be more than the OSTree builder).

That said, @ramcq made the good point on Slack that if the new ref name is stored in `ostree.endoflife-rebase`, old versions of eos-updater will ignore the bit that says ‘reboot into this before proceeding’ and will greedily update past it. There was a bug in eos-updater meaning it wouldn’t follow two `eos.endoflife-rebase` commits in a row, so we could potentially hack our way past this problem that way; but I’m losing the will to fight this battle.

If upstream OSTree does standardise a way of doing checkpoint commits in future, they’re likely to base it around `ostree.endoflife-rebase`. If so, we can always add a checkpoint using our mechanism, then switch to the OSTree mechanism. So there is a graceful transition away from this.

Let’s go with `eos.upgrade-ref-for-deployment`. (Though maybe call it something shorter like `eos.checkpoint-target`?)

smspillaz replied:
I'm happy to go with `eos.checkpoint-target`

pwithnall commented on tests/eos-test-utils.h:
```diff
@@ -99,6 +99,11 @@ struct _EosTestSubserver
    * are a pointer array of SimpleFile instances. */
   GHashTable *additional_files_for_commit;
+
+  /* Mapping from commits numbers to hashtables of metadata string
```
s/commits/commit/?

pwithnall commented on tests/ostree-spawn.c:
```diff
@@ -112,6 +114,36 @@ ostree_init (GFile *repo,
+
+  while (g_hash_table_iter_next (&iter, &key, &value))
+    {
+      gchar *formatted_metadata_string = g_strdup_printf ("%s=%s",
```
`g_autofree`

pwithnall replied:
Perhaps `g_autofree` + `g_steal_pointer()` somewhere appropriate then? I haven’t looked at the code to see what would fit, but I would much prefer to make all ownership transfers explicit using those two annotations, rather than have them implicit in the code (and thus easier to break when refactoring).

pwithnall commented on tests/ostree-spawn.c:
```diff
@@ -135,11 +168,20 @@ ostree_commit (GFile *repo,
+  g_autoptr(GArray) cmd_args = g_array_sized_new (FALSE, TRUE, sizeof (CmdArg), 8);
+  g_autoptr(GPtrArray) formatted_cmd_args_membuf = g_ptr_array_new_with_free_func (g_free);
+  memcpy (cmd_args->data, initial_args, sizeof(initial_args));
+  cmd_args->len = G_N_ELEMENTS (initial_args);
```
This should probably be `g_array_append_vals()`, right?

smspillaz replied:
That could also work.

smspillaz commented:
Comments addressed. Note that there are still conflicts on this branch - I tried rebasing it on (the recently updated) T16682 and it tended up being a quagmire of merge conflicts. I think a better way forward would be to create a new branch and cherry-pick the updated commits on top of the now-rebased T16682 after it has been squashed down into something a little more manageable.

I'll try doing that tonight.

pwithnall commented on Makefile.am:
```diff
@@ -60,11 +60,33 @@ AM_CPPFLAGS = \
+	eos-updater-flatpak-installer/installer.h \
+	$(NULL)
+enum_template_h_file = libeos-updater-util/eos-updater-enums.h.in
+enum_template_c_file = libeos-updater-util/eos-updater-enums.c.in
```
Drop the `eos-updater-` prefix from the files, and call them `libeos-updater-util/enums.[ch]{,.in}` instead. I’m trying to move to a model where filenames don’t have a namespace prefix, because the directory they’re in provides their namespace. It makes tab-completion easier (and reduces typing for the filenames).

pwithnall commented on Makefile.am:
```diff
@@ -87,13 +110,18 @@ libeos_updater_util_libeos_updater_util_@EUU_API_VERSION@_la_SOURCES = \
 	libeos-updater-util/config.c \
 	libeos-updater-util/config.h \
+	libeos-updater-util/eos-updater-types.c \
+	libeos-updater-util/eos-updater-types.h \
```
Moving `eos-updater-types.[ch]` to `libeos-updater-util` makes sense, but please rename them to drop the prefix. So they should become `libeos-updater-util/types.[ch]`. See my comments about `eos-updater-enums.[ch]` for the reasoning.

pwithnall commented on data/eos-updater-flatpak-installer.service.in:
```diff
@@ -13,6 +13,29 @@ Type=oneshot
+ProtectHome=yes
+ProtectKernelModules=yes
+ProtectKernelTunables=yes
+ProtectSystem=strict
```
`ProtectSystem=strict` is overwritten by `ProtectSystem=no` immediately afterwards. Drop the `strict` one.

pwithnall commented on data/eos-updater-flatpak-installer.service.in:
```diff
@@ -13,6 +13,29 @@ Type=oneshot
 ExecStart=@libexecdir@/eos-updater-flatpak-installer
 
+# Sandboxing
+# FIXME: Enable more of these options once we have systemd > 229
```
You can drop the `FIXME` comment.

pwithnall commented on data/eos-updater-flatpak-installer.service.in:
```diff
@@ -13,6 +13,29 @@ Type=oneshot
 
+# Sandboxing
+# FIXME: Enable more of these options once we have systemd > 229
+Environment=GIO_USE_VFS=local
```
What about `CapabilityBoundingSet` and `RestrictNamespaces`? From looking at `man systemd.exec`, I can’t see any other keys we might want, but please do check yourself!

pwithnall commented on debian/eos-updater.install:
```diff
@@ -1,4 +1,5 @@
 lib/systemd/system/eos-autoupdater.service
+lib/systemd/system/eos-updater-flatpak-installer.service
```
I think we also need to add the flatpak dependency to `debian/control` in the relevant places, unless you’ve done that somewhere else?

pwithnall commented on src/eos-updater-apply.c:
```diff
@@ -73,41 +79,304 @@ get_test_osname (void)
+   * any flatpaks which were deployed */
+  flatpaks_to_deploy = eos_updater_util_flatten_flatpak_ref_actions_table (relevant_flatpak_ref_actions);
+
+  if (!pull_flatpaks (flatpaks_to_deploy, cancellable, error))
```
Hmm, should we not be pulling the flatpaks as part of eos-updater’s fetch step, rather than its apply step?

smspillaz replied:
This is a very good point and something I should have thought of earlier when I was designing this. But ....

I think if we were to do the flatpak pulling into the fetch step, it could introduce a lot more complexity and cause this PR to go from a 4000 line diff to 6 or 7000 lines (with commensurate time taken up in review churn). The reason being that we still need to cleanup the right undeployed flatpaks in both the `fetch` and the `apply` error case and I could see that becoming a bunch of complicated state to manage.

Would you be happy with keeping the flatpak pulling in the apply step for now and looking into doing the pulling in the fetch step in a follow-up?

pwithnall commented on data/eos-updater-flatpak-installer.service.in:
```diff
@@ -0,0 +1,41 @@
+MemoryDenyWriteExecute=yes
+NoNewPrivileges=yes
+PrivateDevices=yes
+PrivateNetwork=no
```
`eos-updater-flatpak-installer` shouldn’t have to access the network, should it? The pulls of all the flatpaks are done by `eos-updater`.

smspillaz replied:
Yup, that makes sense. I basically cargo-culted this file 😅

pwithnall commented on src/eos-updater-fetch.c:
```diff
@@ -295,18 +299,451 @@ content_fetch_old (EosUpdater      *updater,
+                                    GCancellable  *cancellable,
+                                    GError       **error)
+{
+  g_auto(GStrv) override_dirs = g_strsplit(eos_updater_util_flatpak_autoinstall_override_paths (), ";", -1);
```
Nitpick: Missing space before `(` in the `g_strsplit()` call.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+
+      if (existing_actions_file &&
+          existing_actions_file->priority < priority)
+        continue;
```
You could do this check before calling `read_flatpak_ref_actions_from_file()` and avoid parsing the file unnecessarily.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+      architecture_skip_filter_applies (filter_value, is_filtered, error))
+    return TRUE;
+  if (g_strcmp0 (filter_key_name, "architectures") == 0 &&
+           architecture_only_filter_applies (filter_value, is_filtered, error))
```
Nitpick: Funky indentation here and in the other `if` statements in this function.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+static gboolean
+locale_only_filter_applies (JsonNode  *filter_value,
+                                  gboolean  *is_filtered,
+                                  GError   **error)
```
Nitpick: Funky indentation for these arguments.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+                                                   error))
+    return FALSE;
+
+  if (skip_locales_array == NULL)
```
Might as well do this `(skip_locales_array == NULL)` check before calling `get_locales_list_from_flatpak_installation()`, to avoid some wasted work on the error path.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+
+  /* TODO: Right now this returns a only the testing override or NULL,
+   * but we might want to do something a little more clever based on what is
+   * supported by Flatpak in future */
```
Add a link to https://github.com/flatpak/flatpak/issues/1156 in the comment

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+{
+  g_autoptr(FlatpakInstallation) installation = eos_updater_get_flatpak_installation (NULL, error);
+  g_autoptr(GError) local_error = NULL;
+  g_autofree gchar *value = NULL;
```
`local_error` and `value` seem to be unused.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+  const gchar *override_locales = eos_updater_get_envvar_or ("EOS_UPDATER_TEST_UPDATER_OVERRIDE_LOCALES", NULL);
+
+  if (override_locales)
+    return g_strsplit(override_locales, ";", -1);
```
Nitpick: Missing space before `(`.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+  return TRUE;
+}
+
+static gboolean
```
All four of these filtering functions are very structurally similar. You could probably merge them into a single function quite easily (which takes a `GStrv` of strings to match against: either the locale list, or a singleton list of the current architecture; and which also takes a `gboolean` telling it whether the match is inverted).

smspillaz replied:
I was actually thinking this when I wrote these, just wanted to make sure the tests were passing first before I did this refactor ... then I must not have gotten around to it. I'll do that now, since its relatively trivial.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+      return NULL;
+    }
+
+  action_type_str = json_object_get_string_member (object, "action");
```
Unfortunately the `json_object_get_*_member()` methods will hit an assertion failure if no member exists by that name, or if it has the wrong type. You need to use `json_object_get_member()`, check it’s non-`NULL`, check its type, then get its string value. :-(

smspillaz replied:
That's unfortunate. Would a patch to json-glib for `json_object_safe_get_*_member` be accepted?

pwithnall replied:
I would *love* to see that. @ebassi is the person to speak to, since he maintains json-glib.

ebassi replied:
See [this issue](https://gitlab.gnome.org/GNOME/json-glib/issues/4).

Short answer: no, not really.

Longer answer: JSON has no validating schema, and C does not have an "unset" value; a missing object member returning `NULL` or `0` is indistinguishable from an existing object member holding that value.

The usual pattern for GLib-based API would be to have a `json_object_get_<type>_member()` version with a `GError` argument, and let you handle the missing key error condition; I find it distasteful in `GKeyFile` not only because it makes for heavily convoluted error handling code, but also because it will require the returned value to be an out parameter, in order to be handled correctly by language bindings. So I'm not going to accept a patch that does that.

The only suitable replacement would be a Python-esque `json_object_get_<type>_member_with_default()` that allows you to specify the default value in case the member is missing. The API would be quite a mouthful, but I'd consider it the only feasible approach.

pwithnall replied:
OSTree uses out-parameters to return values quite successfully. I’d be happy with either of:
```
gboolean json_object_get_<type>_member (JsonObject *obj, const gchar *member_name, <type> *out_value);
```
which returns `TRUE` iff the member exists and is of `<type>` (and hence has been returned in `out_value`).

*Or*

```
<type> json_object_get_<type>_member_with_default (JsonObject *obj, const gchar *member_name, const <type> default);
```

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+
+  serial64 = json_node_get_int (serial_node);
+
+  if (serial64 > G_MAXINT32 || serial64 < G_MININT32)
```
This restriction isn’t enforced in the `.schema.json` file. (It makes sense, but it should be added to the JSON Schema.)

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+  if (!parse_flatpak_ref_from_entry (entry, &app_name, &kind, error))
+    return NULL;
+
+  collection_id = json_object_get_string_member (entry, "collection-id");
```
See comment below: use of `json_object_get_*_member()` will cause assertion failures if that member doesn’t exist or has the wrong type.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+      return FALSE;
+    }
+
+  ref_kind_str = json_object_get_string_member (entry, "ref-kind");
```
See comment below: use of `json_object_get_*_member()` will cause assertion failures if that member doesn’t exist or has the wrong type.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+                                      GError                                   **error)
+{
+  GEnumClass *enum_class = g_type_class_ref (EOS_TYPE_UPDATER_UTIL_FLATPAK_REMOTE_REF_ACTION_TYPE);
+  GEnumValue *enum_value = g_enum_get_value_by_nick (enum_class, action);
```
I’d prefer using explicit hard-coded strings here, rather than the nicknames from the enum definition. It’s more explicit, and less likely to magically break if we refactor the code later and change the enum.

smspillaz replied:
@cosimoc Given that you were the one who recommended doing it this way, did you have any thoughts given the above ^ ?

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+{
+  FlatpakRemoteRefAction *action = g_slice_new0 (FlatpakRemoteRefAction);
+
+  action->ref_cnt = 1;
```
Call it `ref_count` please; no nd t abbrvt so harshly.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+FlatpakRemoteRefAction *
+flatpak_remote_ref_action_ref (FlatpakRemoteRefAction *action)
+{
+  ++action->ref_cnt;
```
Add preconditions:
```
g_return_val_if_fail (action->ref_count > 0, NULL);
g_return_val_if_fail (action->ref_count < G_MAXUINT, NULL);
```
(these are the standard preconditions for any `ref()` function)

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+void
+flatpak_remote_ref_action_unref (FlatpakRemoteRefAction *action)
+{
+  if (--action->ref_cnt != 0)
```
Add preconditions:
```
g_return_if_fail (action->ref_count > 0);
```
(this is the standard precondition for `unref()` functions)

smspillaz replied:
I'll add these given that they are the standard, but given that `action->ref_count` is `guint` I don't think they will serve much use?

pwithnall replied:
Helps in the double-unref case, when `action->ref_count == 0` at the start of the second call.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+                                                            flatpak_ref_equal,
+                                                            g_object_unref,
+                                                            (GDestroyNotify) flatpak_remote_ref_action_unref);
+  GPtrArray *squashed_ref_actions = NULL;
```
`g_autoptr()`

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+      FlatpakRemoteRefAction *action = g_ptr_array_index (ref_actions, i);
+      g_hash_table_replace (hash_table,
+                            g_object_ref (action->ref),
+                            flatpak_remote_ref_action_ref (action));
```
This algorithm is going to need a bit of work for `upgrade` support, since [install X, upgrade X] squashes to [install X], and [uninstall X, upgrade X] squashes to [uninstall X].

smspillaz replied:
I'll do the work to support `upgrade` once we've worked out the rest of these review comments

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+    g_ptr_array_add (squashed_ref_actions,
+                     flatpak_remote_ref_action_ref (value));
+
+  g_ptr_array_sort (squashed_ref_actions, sort_flatpak_remote_ref_actions);
```
Should we be sorting here? By this point, the serial numbers don’t form a total order, since we’re dealing with entries from multiple files (serial number domains). I don’t think you can take a round-trip through the hash table: you’d need to build `squashed_ref_actions` from the original `ref_actions` with a lookup in the hash table for each entry to check whether it’s the latest for that ref (and hence whether to include it).

smspillaz replied:
See my comment below - the sorting here is what should be done based on the current design doc.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+      g_hash_table_insert (squashed_ref_actions_table,
+                           g_strdup (key),
+                           squashed_remote_ref_actions);
+    }
```
We need to squash the refs once, across all the serial number domains, not once per domain.

smspillaz replied:
Hmm, my understanding of the design doc was that we wanted to do this:

 - read in all serial domains
 - squash each one
 - concatenate all the squashed lists in lexicographical order
 - squash again

Based on the "Order of applying entries from autoinstall lists" section (see "Worked Example")

pwithnall replied:
I don’t think the first compression step (of individual files) is necessary. Are there any situations in which the final list of actions would differ depending on whether the first compression step was done?

I removed the first compression step from §(Worked example) mid-last-week. Sorry I didn’t point it out at the time. (You should probably re-check the worked example matches your understanding now.) (A diff view in Google Docs and an easy way to see ‘changes since I last viewed this document’ would be useful.)

smspillaz replied:
Off the top of my head, I don't *think* there would be any difference. @cosimoc  was the one who suggested the approach of compression twice, perhaps he can lend some insight?

smspillaz replied:
Ah damn, okay, we do need to do the initial squash.

It actually isn't so important for the step where the flatpaks are pulled, but it does matter for the step where the flatpaks are installed by `eos-updater-flatpak-installer`, since in that program, we do not assemble all actions into a single pointer array. This is because whilst we are running the installations, we need to keep track to see if any failed in a domain and if so update the counter accordingly. Therefore, the squash step needs to happen on each domain but not on the whole lot.

pwithnall replied:
Hmm. I’d prefer to apply the squash to the whole lot, since that could eliminate some actions which would cause apps to be downloaded unnecessarily (using the user’s bandwidth).

Three possible solutions to this:
 - Don’t update the stored serial number for that domain. This has the downsides we discussed before on the design document (§(Downloading and deploying flatpaks on OS upgrade)).
 - Re-parse the JSON files on failure to find the ID before the one which failed.
 - Store the ID from the previous action in a new field in our `struct` for storing a parsed action, and set the stored serial number to that on failure. (Better than option 2 since it doesn’t require re-parsing.)

I think I prefer the third option, and it should be fairly simple to implement, I think?

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+ * to a counter indicating how up to date we are on the system for that
+ * file */
+GHashTable *
+eos_updater_util_remote_ref_actions_to_expected_progresses (GHashTable *ref_actions_table)
```
Is this function used? From some quick grepping it doesn’t seem to be. Unless it’s used in the tests, which I haven’t loaded for this review.

smspillaz replied:
Hum, seems like it isn't used anymore. I'll remove it.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+      if (ref_actions->len == 0)
+        {
+          g_warning ("Remote ref actions file %s has no actions, that's odd...",
+                     key);
```
No need for a warning here: having no entries in an autoinstall file is perfectly acceptable. This could be a `g_debug()` if you want.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+  /* Enumerate each section. The section name is the path to the file */
+  groups_iter = groups = g_key_file_get_groups (state_key_file, NULL);
+
+  while (*groups_iter != NULL)
```
Restructure this as a `for` loop to make the termination condition and iteration more obvious.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+      g_autofree gchar *progress_string = NULL;
+      guint64 progress;
+
+      /* We need to use g_key_file_get_value here to guard against errors */
```
To guard against what kind of errors? Why not `g_key_file_get_integer()`?

smspillaz replied:
See https://developer.gnome.org/glib/stable/glib-Key-value-file-parser.html#g-key-file-get-integer - "If key cannot be found then 0 is returned and error is set to G_KEY_FILE_ERROR_KEY_NOT_FOUND. Likewise, if the value associated with key cannot be interpreted as an integer, or is out of range for a gint, then 0 is returned and error is set to G_KEY_FILE_ERROR_INVALID_VALUE.".

That said, we could drop this if we did not allow "0" to be in-range.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+                                              error);
+
+      if (progress_string == NULL)
+        return NULL;
```
If `progress_string == NULL`, should we not just not insert an entry for `source_path` in the hash table, and carry on? I think it would be fine (if unexpected) for the progress file to have an empty section — just treat it like the section doesn’t exist.

smspillaz replied:
I think this is meant to be an invariant - if there is a section in the keyfile for that filename, then it should have a `Progress` key.

pwithnall replied:
If you want. I generally put invariants into user-facing code like this only if it makes things simpler for me, and I’m not sure that’s the case here.

As long as whatever you do is documented (brief code comment is fine) I’ll be happy.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+{
+  gpointer key, value;
+  GHashTableIter iter;
+  GString *string = g_string_new ("");
```
`g_autoptr()`

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+        }
+    }
+
+  return g_string_free (string, FALSE);
```
`return g_string_free (g_steal_pointer (&string), FALSE);`

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -0,0 +1,1411 @@
+{
+  gpointer key, value;
+  GHashTableIter iter;
+  GString *string = g_string_new ("Action application progresses:\n");
```
`g_autoptr()`

pwithnall commented on libeos-updater-util/flatpak.h:
```diff
@@ -0,0 +1,102 @@
+G_BEGIN_DECLS
+
+#define FLATPAKS_IN_OSTREE_PRIORITY 100
+#define FLATPAKS_IN_OVERRIDE_DIR_PRIORITY 0
```
These could do with some documentation.

pwithnall commented on libeos-updater-util/flatpak.h:
```diff
@@ -0,0 +1,102 @@
+} EosUpdaterUtilFlatpakRemoteRefActionType;
+
+typedef struct {
+  unsigned int                             ref_cnt;
```
`guint` for consistency with everything else please

pwithnall commented on libeos-updater-util/util.c:
```diff
@@ -673,3 +673,127 @@ eos_string_to_unsigned (const gchar  *str,
+{
+  const gchar *override_path = g_getenv ("EOS_UPDATER_TEST_FLATPAK_INSTALLATION_DIR");
+
+  if (override_path)
```
Nitpick: `override_path != NULL`

pwithnall commented on src/eos-updater-poll-common.c:
```diff
@@ -312,6 +310,111 @@ get_booted_refspec (gchar               **booted_refspec,
+      return TRUE;
+    }
+
+  *out_refspec_to_upgrade_from_deployment = g_variant_dup_string (ref_for_deployment_variant, NULL);
```
Before we return this, we should check that it’s a valid refspec, just in case (for example), `eos.checkpoint-target` gets set to an empty string in future, or something.

pwithnall commented on src/eos-updater-poll-common.c:
```diff
@@ -796,6 +899,24 @@ run_fetchers (EosMetadataFetchData *fetch_data,
+      else
+        {
+          g_message ("Failed to get metrics: %s", metrics_error->message);
+        }
```
This chunk looks like a dodgy merge: the metrics are reported by the `metrics_report_successful_poll() call immediately below.

smspillaz replied:
Yeah ... I did think that something was amiss when this came up in the merge conflicts whilst rebasing. Thanks for spotting.

pwithnall commented on src/eos-updater-poll-common.c:
```diff
@@ -312,6 +310,111 @@ get_booted_refspec (gchar               **booted_refspec,
+}
+
+gboolean
+get_refspec_to_upgrade_on (gchar               **refspec_to_upgrade_on,
```
Note to self: I need to check how this integrates with `fetch_latest_commit()` and `parse_latest_commit()`.

smspillaz commented on src/Makefile.am:
```diff
@@ -73,13 +73,13 @@ common_cppflags = \
 eos_autoupdater_CFLAGS = $(common_cflags) $(EOS_AUTOUPDATER_CFLAGS)
 eos_autoupdater_LDFLAGS = $(WARN_LDFLAGS)
-eos_autoupdater_LDADD = $(CODE_COVERAGE_LIBS) $(GIO_LIBS) $(dbuslib) $(top_builddir)/libeos-updater-util/libeos-updater-util-@EUU_API_VERSION@.la $(EOS_AUTOUPDATER_LIBS)
+eos_autoupdater_LDADD = $(CODE_COVERAGE_LIBS) $(FLATPAK_LIBS) $(GIO_LIBS) $(dbuslib) $(top_builddir)/libeos-updater-util/libeos-updater-util-@EUU_API_VERSION@.la $(EOS_AUTOUPDATER_LIBS)
```
If I remember correctly, this was something to do with `eos-autoupdater` linking to `libeos-updater-util` which needed to link to `libflatpak` (and so due to DSO rules, `eos-autoupdater` also needed to be linked to `libflatpak`). I can remove it and double check, though.

smspillaz replied:
Nope, looks like this isn't necessary. Removed.

smspillaz commented on src/eos-updater-fetch.c:
```diff
@@ -295,18 +299,451 @@ content_fetch_old (EosUpdater      *updater,
+
+  g_main_context_push_thread_default (context);
+
+  if (!ostree_repo_open (repo, cancellable, error))
```
Oh wow, good catch, I must have completely forgotten about that.

(Wishlist: Would be nice if there was an RAII-like `GMainContextThreadDefault` object that we could use with `g_autoptr`)

pwithnall replied:
The wishlist can happen if you submit a patch upstream: I’ll review it pronto. :-)

smspillaz commented on src/eos-updater-fetch.c:
```diff
@@ -295,18 +299,451 @@ content_fetch_old (EosUpdater      *updater,
+}
+
+static GFile *
+inspect_directory_in_ostree_repo (OstreeRepo    *repo,
```
I'll move it out into a helper library, but I'm not sure how we will unit test something like this without essentially doing what the acceptance tests are already doing (I don't think it's possible to make an in-memory `OstreeRepo`, is it?)

pwithnall replied:
It’s not possible to have an in-memory `OstreeRepo`. I think in this case the main advantage comes from being able to expose more internal state which you can check in the test code, rather than essentially summarising _everything_ in the exit status of a program.

If possible, it would be good to eliminate some environment variables this way. Environment variables are almost always a hack to expose/change internal state from somewhere it shouldn’t be exposed.

smspillaz replied:
Eliminating the environment variables would be nice, but I don't think we can do that without the expense of not being able to do end-to-end testing - they're kind of a necessary evil for that.

(I've refactored things a little so that at least for the parts we expose to the unit tests, we're not reliant on environment variables).

smspillaz commented on src/eos-updater-fetch.c:
```diff
@@ -295,18 +299,451 @@ content_fetch_old (EosUpdater      *updater,
+  if (!ostree_repo_open (repo, cancellable, error))
+    return NULL;
+
+  ostree_repo_find_remotes_async (repo,
```
Hmmm, I'm confused. I thought the reason why we wanted to specify a collection ID was that in flatpak land, `flatpak_remote_name == ostree_remote_name` and this was a bad thing. Are you saying that we can just use a flatpak remote name directly and that will be translated into a collection ID internally? What is the point of making all the configuration files take collection ID's in that case?

smspillaz replied:
In the meantime I'll reimplement using the above suggestion, thanks.

pwithnall replied:
`flatpak_remote_name == ostree_remote_name` always holds. We can use a flatpak remote name directly and that will be translated into a collection ID internally, indeed. There is no immediate need for the configuration files to use collection IDs: this could all work just as well if they used remote names. Over time, I would like things to move to using collection IDs (since they’re more globally unique than remote names), so the idea here was to use them from the start to avoid the need for a checkpoint in future to add them.

Using collection IDs here does mean that we can’t add anything to the install list until https://phabricator.endlessm.com/T19363 and https://phabricator.endlessm.com/T19487 have been fixed. These tasks add the `collection-id=` keys to the repo config files on new and already-installed images.

Since I want to be cautious about enabling P2P updates for everyone, then perhaps we _should_ use remote names in the autoinstall list (but *also* support collection IDs, so that we don’t need a checkpoint to add support for them in future). What do you think?

smspillaz commented:
Okay, most of the comments should now be addressed. The only one that isn't is about moving things into a helper library - I figure it will cause far less rebase pain if we do that after we're happy with how everything else works.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -1199,31 +1150,22 @@ eos_updater_util_flatpak_ref_action_application_progress_in_state_path (GCancell
-        return NULL;
-
-      if (!parse_monotonic_counter (progress_string, &progress, error))
+      if (progress == 0)
```
I think you need to check the `error` here. At the moment it’s possible for `Progress=0` to be set in the key file, and this code will return `NULL` with no error set.

smspillaz replied:
One of the invariants set earlier was that `Progress > 0`. That said, it wouldn't hurt to check.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -289,14 +295,7 @@ flatpak_remote_ref_action_from_json_node (JsonNode *node,
-                   "Expected a 'serial' member in the autoinstall spec (at %s)", node_str);
-      return NULL;
-    }
+    return NULL;
```
This will result in returning `NULL` but with no error set, I think. Is that OK?

smspillaz replied:
Ah yes, we'll need to set the error - must have overlooked it.

pwithnall commented on libeos-updater-util/flatpak.c:
```diff
@@ -906,9 +925,39 @@ squash_ref_actions_ptr_array (GPtrArray *ref_actions)
+        }
+      else if (action->type == EUU_FLATPAK_REMOTE_REF_ACTION_UPDATE)
+        {
+          if (existing_action_for_ref->type == EUU_FLATPAK_REMOTE_REF_ACTION_UNINSTALL)
```
This block implements `[uninstall X, update X]` → `[update X]`, right? I don’t think that’s right. It should compress to `[uninstall X]`, because the update entry does nothing if the app is not installed. I think it has to be that way, because if that list is split in two and applied in two separate runs of the flatpak updater (for example, the first entry exists on day X, then an update happens and the second entry is added on day X+ 5), the outcome is `[uninstall X]`, not `[update X]`.

smspillaz replied:
Ah yes, that's a good point. I'll change it around.

smspillaz commented:
Okay, did a bunch more work on this:

 - Addressed review comments
 - Added test cases specified in the phabricator ticket
 - Refactored things a little to support those test cases
 - Required that JSON files specify both "remote" and "collection-id" for now, as discussed
 - Dropped the "initial compression" step.

pwithnall commented on src/eos-updater-fetch.c:
```diff
@@ -295,18 +299,460 @@ content_fetch_old (EosUpdater      *updater,
+}
+
+static gboolean
+content_fetch (EosUpdater      *updater,
```
Note for future branches: it’s a lot easier to review stuff like this if the branch is structured as a series of commits which, separately: refactor the existing code without changing its behaviour (e.g. pull `content_fetch()` out of `content_fetch_task()`), add the new code (e.g. `flatpak.c`), and then add calls to the new code into the old code (e.g. the `prepare_flatpaks_to_deploy()` call below). The third step would ideally be as minimal as possible. Then a reviewer who cares about ensuring the existing code continues to work only has to review the first commit in detail — and that first and second commit can potentially be merged in advance of the third one, since they can easily be guaranteed to not break anything.

(I don’t want this to be changed in this branch now; that’s too much rebase faff.)

pwithnall replied:
I just pushed the `commit_fetch()`/`commit_fetch_task()` split to master, so you’re going to get some minor rebase problems here.

pwithnall commented on src/eos-updater-poll-common.c:
```diff
@@ -796,6 +906,8 @@ run_fetchers (EosMetadataFetchData *fetch_data,
     {
       EosUpdateInfo *latest_update = NULL;
+      g_autofree gchar *booted_ref = NULL;
+      g_autoptr(GError) metrics_error = NULL;
```
These two variables seem to have been added spuriously.

pwithnall commented on src/eos-updater-poll-common.c:
```diff
@@ -312,6 +310,118 @@ get_booted_refspec (gchar               **booted_refspec,
+
+  /* We'll need the repo to get the collection-id */
+  if (!ostree_sysroot_get_repo (sysroot, &repo, NULL, error))
+    return FALSE;
```
This could move to below the `/* Nothing here */` bailout.

pwithnall commented on src/eos-updater-poll-common.c:
```diff
@@ -312,6 +310,118 @@ get_booted_refspec (gchar               **booted_refspec,
+  g_autoptr(GVariant) metadata = NULL;
+  g_autoptr(GVariant) ref_for_deployment_variant = NULL;
+  g_autofree gchar *refspec_for_deployment = NULL;
+  OstreeRepo *repo = NULL;
```
`repo` should be `g_autoptr()`.

pwithnall commented on src/eos-updater-poll-common.c:
```diff
@@ -312,6 +310,118 @@ get_booted_refspec (gchar               **booted_refspec,
+
+  refspec_for_deployment = g_variant_dup_string (ref_for_deployment_variant, NULL);
+
+  if (refspec_for_deployment != NULL &&
```
I’m pretty sure `refspec_for_deployment` can’t be `NULL` here, since `g_variant_dup_string()` should never return a `NULL` string (it can’t fail, and `GVariant`s can’t contain `NULL` strings).

pwithnall commented on configure.ac:
```diff
@@ -109,21 +120,31 @@ AS_IF([test "$GDBUS_CODEGEN" = ""],[
-	AC_MSG_ERROR([ostree is required for tests])
+AC_PATH_PROG([OSTREE_BINARY],[ostree])
+AS_IF([test "$OSTREE_BINARY" = ""],[
+	AC_MSG_ERROR([ostree command line tool is required for tests])
```
When I added comments about this before, I just meant for you to change the error message, not s/`OSTREE`/`OSTREE_BINARY`/. I think that latter change makes things more confusing, given that the convention for program names in autotools is to be the program name in uppercase, i.e. `$(OSTREE)`, not `$(OSTREE_WITH_SOME_SUFFIX)`.

I merged an edited version of this commit to `master` already, which just changes the existing error messages, so you’ll get a minor rebase conflict on this. Could you drop the `_BINARY` changes at the same time please?

pwithnall commented on src/eos-updater-fetch.c:
```diff
@@ -295,18 +299,460 @@ content_fetch_old (EosUpdater      *updater,
+}
+
+static GHashTable *
+flatpak_ref_actions_for_commit (OstreeRepo    *repo,
```
Can this be moved out to the utility library please? Then it can be easily tested on a temporary `OstreeRepo` without having to queue up `eos-updater` with a fetch operation for that repo.

pwithnall commented:
I’ve pushed a couple of fixup commits to the branch, which fix building with `builddir` ≠ `srcdir`.

pwithnall commented on src/eos-updater-poll-common.c:
```diff
@@ -312,6 +310,118 @@ get_booted_refspec (gchar               **booted_refspec,
+  metadata = g_variant_get_child_value (commit, 0);
+  ref_for_deployment_variant = g_variant_lookup_value (metadata,
+                                                       "eos.checkpoint-target",
+                                                       G_VARIANT_TYPE_STRING);
```
Since you’re going to have to rebase and squash everything here anyway, would it be possible to split the checkpoint stuff (the changes here in `poll-common.c` and the new tests for it) into a new PR, which I can review and merge separately (probably before this PR gets merged). That should help to unblock things — generally splitting large branches like this up into easily-mergeable groundwork PRs, followed by a larger PR for the new feature where its interface with the existing code is minimal and easily verifiable to be correct, is good. The checkpoint stuff counts as groundwork, as does splitting `test-common` out.

wrt rebasing everything else once you’ve split those out, I think it might be easiest if you just squashed everything together, then split it out again a bit as necessary (using `git reset HEAD~; while true; do git add -p; git commit; done`). It seems like a lot of the commits in the branch are thinly-veiled fixups, and trying to disentangle them all and squash them in to the right commits individually is going to be too hard. Given that you’re ultimately not going to end up with many commits on the branch (since the feature is pretty monolithic once the groundwork has been split out), it’s probably easiest to start from a single monolithic commit and break it down a bit.

I hope that makes sense. If not, poke me and we can discuss it. I should be able to find time to do the rebase/rearrangement of the branch myself if you think that would help or be instructive. I don’t want to step on your toes with it if you have a plan already though. :-)

smspillaz commented:
@pwithnall thanks for the reviews and fixups!

With respect to rebasing ... I think what I might do is fix up the remaining identified points on this branch, then start three new branches

1. Split out all the tests stuff into tests-common
2. Checkpoint commits
3. Flatpak auto-installation

They do each step on each other a little bit, but I'm okay to bear that pain as a tradeoff for reducing the diff size and unblocking.

pwithnall commented:
Sure, that works for me. Within each branch, the more you can split out groundwork commits from other commits, the better; as I can cherry-pick from within branches as well. The three branches will step on each other a bit, but that shouldn’t be a problem as long as they’re reviewed and applied in the intended order. I’m not expecting drastic review comments on any of them, which will minimise rebase collisions.

If you could make sure that each branch passes the unit tests when it’s applied, and make sure that eos-updater still compiles between each commit (I use `git for-each master.. make all`, with `git-for-each` from [git-branch-tools](https://github.com/whot/git-branch-tools)), that acts as a bit of assurance that changes have been split out into separate commits properly, and that cherry-picking one of the commits will not cause problems. Thanks!
